### PR TITLE
ig: Switch to use container name from runtime and show runtime metadata in gadgets 

### PR DESCRIPTION
**Pull request by blanquicet**
# ig: Switch to use container name from runtime and show runtime metadata in gadgets

This PR is composed of three main changes:

1. Creating sub-structs for k8s and runtime metadata in the events and container structs. This change applies for `ig` and `kubectl-gadget`. The JSON output is now like this:

      ```bash
      $ sudo ig trace exec
      [...]
      {
        "runtime": {
          "runtime": "cri-o",
          "containerId": "71928cf10b3f72d9bbf72d5b0948568ebac78221258c4ab9917bc0fce98320cf",
          "container": "k8s_kube-proxy_kube-proxy-plq7g_kube-system_94247867-3231-46ab-a578-2608b227b0c1_0"
        },
        "k8s": {
          "namespace": "kube-system",
          "pod": "kube-proxy-plq7g",
          "container": "kube-proxy"
        },
        "timestamp": 1686083037200509658,
        "type": "normal",
        "mountnsid": 4026534463,
        "pid": 587914,
        "ppid": 20979,
        "comm": "iptables",
        "args": [
          "/usr/sbin/iptables",
          "-w",
          "5",
          "-W",
          "100000",
          "-S",
          "KUBE-PROXY-CANARY",
          "-t",
          "mangle"
        ],
        "uid": 0,
        "gid": 0,
        "loginuid": 4294967295,
        "sessionid": 4294967295
      }
      [...]
      ```
      
      Regarding the columns output format, it doesn't change for `kubectl-gadget, while it changes a bit for `ig` because of the `Runtime` prefix that this PR adds to the runtime metadata fields. Please take a look at item (3) for more details.

2. Adding runtime metadata to events in `ig` and `kubectl-gadget`:
    - **Container ID**: It was already available but not under the runtime metadata sub-struct.
    - **Container Runtime Name**: docker, containerd, CRI-O or Podman.
    - **Container Name**: The name given by the container runtime. Notice it might be different than the name Kubernetes gave. In addition, this field will not be available for `kubectl-gadget` when using Docker as runtime in k8s. Such a case will be managed in a future PR.

3. Removing [this workaround](https://github.com/inspektor-gadget/inspektor-gadget/blob/dd7922590b9b8e9d433e412ff5e9dd627490946c/pkg/container-collection/options.go#L55-L58) and enriching the containers (and thus the events) with both container names: by runtime and k8s. It means that now `ig` will show the name given by the runtime (as it should have been since the beginning) instead of the k8s one in the container column. In addition, we added the `RUNTIME` prefix for all the runtime metadata. Therefore, the new `ig` UX looks like this:

      ```bash
      $ ig trace exec
      RUNTIME.R… RUNTIME.CONTAINERID RUNTIME.CONTAINER                     PID        PPID       COMM             RET ARGS
      containerd cd8ce885c115adbc…   nsenter                               149151     5786       check_scheduled  0   /etc/node-problem-detector.d/plugin/check_schedul…
      containerd cd8ce885c115adbc…   nsenter                               149152     149151     curl             0   /usr/bin/curl -s -H Metadata:true --noproxy * htt…
      containerd cd8ce885c115adbc…   nsenter                               149155     149153     jq               0   /usr/bin/jq --arg vm aks-nodepool1-28093845-vmss_…
      containerd cd8ce885c115adbc…   nsenter                               149158     149156     jq               0   /usr/bin/jq length
      containerd cd8ce885c115adbc…   nsenter                               149159     5786       check_preempt.s  0   /etc/node-problem-detector.d/plugin/check_preempt…
      containerd cd8ce885c115adbc…   nsenter                               149160     149159     check_scheduled  0   /etc/node-problem-detector.d/plugin/check_schedul…
      containerd cd8ce885c115adbc…   nsenter                               149161     149160     curl             0   /usr/bin/curl -m 24 -H Metadata:true --noproxy * …
      containerd cd8ce885c115adbc…   nsenter                               149164     149162     jq               0   /usr/bin/jq --arg vm aks-nodepool1-28093845-vmss_…
      containerd cd8ce885c115adbc…   nsenter                               149167     149165     jq               0   /usr/bin/jq length
      containerd cd8ce885c115adbc…   nsenter                               149168     2716       runc             0   /usr/bin/runc --root /run/containerd/runc/k8s.io …
      containerd cd8ce885c115adbc…   nsenter                               149175     149168     exe              0   /proc/self/exe init
      containerd fada3094230e1cd5…   node-driver-registrar                 149178     149168     csi-node-driver  0   /csi-node-driver-registrar --kubelet-registration…
      ```

Related https://github.com/inspektor-gadget/inspektor-gadget/issues/737

### TODO (Before merging): 

- [x] Fix crash when comparing `containers.ownerReference` when `list-containers` tests fail.
- [x] Change `K8sMetadata` and `RuntimeMetadata` variable names to `K8s` and `Runtime`. 
- [x] Change `Runtime.ID` to `Runtime.ContainerID`
- [x] Is `Runtime` the correct word for high-level container runtime metadata? 
  - We didn't find a better word.
- [x] Should it be `ContainerName` instead of `Container`?
  - Using ContainerName
- [x] Should it be `Runtime.Name` instead of `Runtime.Runtime`?
  - Using Runtime.RuntimeName.
- [x] Update integration tests
- [x] Allow getting sortable fields with `columnTags` so that CR documentation is not generated with runtime metadata fields (At least until we can guarantee such data will always be available in `kubectl-gadget`).
    - I updated the doc as CR code is going to change soon and I wanted to avoid implementing something that might get removed with coming changes. 

### TODO (After merging, another PR):

- [ ] Ensure runtime metadata is available for `kubectl-gadget` and verify them on integration tests.
- [ ] Add flag to show k8s columns in all `ig` commands with the prefix `K8S`.
- [ ] Ensure Runtime information is guaranteed in ig-k8s (`kubectl-gadget`). For instance, the container name for CRI-O can be retrieved from the OCI-Annotations or from the container-hook.
- [ ] `Runtime` and `K8s` struct within `CommonData` should be pointers, see https://github.com/inspektor-gadget/inspektor-gadget/pull/1702#discussion_r1258916918.
- [ ] Rename `GetContainer()`, `GetNamespace()`, etc. to specify it is about k8s information.
- [ ] Update proto Container definition to be aligned with container-collection/Container definition, see https://github.com/inspektor-gadget/inspektor-gadget/pull/1702#discussion_r1258966121.

### Open questions (For future PR):

- Do we want to add the `K8S` prefix to `kubeclt-gadget`? If we don't, we must find how to make it different for `ig` and `kubectl-gadget`?

### CODE REVIEW

File: integration/helpers.go
Comment (mqasimsarfraz) [inline integration/helpers.go:1]: I see we are still seeing panic related to `cmp.Diff` :( The `IgnoreUnexported` doesn't work recursively so I believe we will have to add `containercollection.K8sMetadata` [here](https://github.com/inspektor-gadget/inspektor-gadget/blob/main/integration/helpers.go#L33) as well.

**Review by mauriciovasquezbernal — COMMENTED** [review]
> Should it be ContainerName instead of Container?

I'm more for having the Name suffix in them.  `ContainerName`, `PodName`, `RuntimeName`, etc. 

> Should it be Runtime.Name instead of Runtime.Runtime?

I think it should be RuntimeName to be totally clear.

Besides few small comments it's looking very good. I wasn't able to check all the changes line by line, but since most of them are renaming and code reorganization I think it's fine.

File: docs/ig.md
Comment (blanquicet) [reply docs/ig.md:17]: Probably a problem while rebasing.

File: pkg/types/types.go
Comment (blanquicet) [reply pkg/types/types.go:112]: We are using CamelCase format, e.g., [`CgroupID`](https://github.com/blanquicet/inspektor-gadget/blob/4e624e6a8ca199144e139d6c63ad8d9baaef933d/pkg/container-collection/containers.go#L56).

**Comment by blanquicet** [comment]
> > Should it be ContainerName instead of Container?
> 
> I'm more for having the Name suffix in them. `ContainerName`, `PodName`, `RuntimeName`, etc.

@mauriciovasquezbernal also in the columns or just for the variable names? Because in that case, it will require updating all the guides/gifs.

**Comment by blanquicet** [comment]
I implemented all the missing tasks, rebased it, and the CI is now green 🥳 !

File: pkg/container-utils/containerutils.go
Comment (mqasimsarfraz) [inline pkg/container-utils/containerutils.go:39]: nit: `types.RuntimeNameDocker.String()`

File: pkg/container-utils/cri/cri.go
Comment (mqasimsarfraz) [inline pkg/container-utils/cri/cri.go:356]: How about using annotation `io.kubernetes.cri-o.Name` for it? You can see it in [workflow run](https://github.com/inspektor-gadget/inspektor-gadget/actions/runs/5486300137/jobs/9996327362#step:7:27193) here. I see it listed as future PR so fine handling it later.

**Review by mqasimsarfraz — APPROVED** [review]
Looking Ready! Please wait for another approval. Making CI green on such changes is always challenging, great work! :partying_face: 

I am thinking if we should clarify, if runtime or k8s container name will be used with `--containername` ?

File: integration/helpers.go
Comment (blanquicet) [reply integration/helpers.go:189]: Yep. It is because it is an option for `BuildCommonData`, which mainly creates a `CommonData` with k8s metadata:

https://github.com/inspektor-gadget/inspektor-gadget/blob/a39e7b00cc2d114259f3d9121aa4232ac7f6e407/integration/helpers.go#L196-L212

Instead, in the non-k8s tests, we manually create the `CommonData`. For example:

https://github.com/inspektor-gadget/inspektor-gadget/blob/a39e7b00cc2d114259f3d9121aa4232ac7f6e407/integration/ig/non-k8s/trace_dns_test.go#L76-L84

I agree `BuildCommonData` could be generalized using the options pattern to work for both cases, but, IMO, we can manage it in another PR. WDYT?

File: integration/inspektor-gadget/advise_networkpolicy_test.go
Comment (blanquicet) [reply integration/inspektor-gadget/advise_networkpolicy_test.go:91]: Yep, I would like first to study and understand the different container enrichers (PodInformer, OCI-annotations, etc), create some tests for them independently and then request the runtime information here. It is because [ig-k8s doesn't directly use the Runtime enricher](https://github.com/inspektor-gadget/inspektor-gadget/blob/a39e7b00cc2d114259f3d9121aa4232ac7f6e407/pkg/gadgettracermanager/gadgettracermanager.go#L251-L320), so I want to be sure the runtime information can always be guaranteed, either by adding the Runtime enricher to ig-k8s or to ensuring the same information can be retrieved by other means in the k8s context.

I'm already working on this. I generalized the TODO for future PRs from " Improve how Runtime.Container is enriched for CRI-O." to "Ensure Runtime information is guaranteed in ig-k8s (`kubectl-gadget`)".

File: pkg/container-utils/cri/cri.go
Comment (blanquicet) [reply pkg/container-utils/cri/cri.go:356]: Yes, I will do that in the third TODO I left for the future PR. I'm also getting it from the container-hook but it is not valid for initial containers. In general, we need to document and test container enrichers as we need to know exactly what elements is each enricher providing. I'll try to work on that on this sprint.

**Comment by blanquicet** [comment]
> I am thinking if we should clarify, if runtime or k8s container name will be used with --containername ?
>
Oops, you are right. Probably it should always be `--runtime.containername | -rc` or `--k8s.containername | -kc` (Not available yet) in `ig`. But then what would we do in `kubectl-gadget` when we also make the runtimes columns available there? At that point, we will have the same problem.

File: integration/helpers.go
Comment (mqasimsarfraz) [reply integration/helpers.go:189]: > but, IMO, we can manage it in another PR. WDYT?

Yeah, We can handle such fixups in seperate PRs.

File: pkg/container-utils/cri/cri.go
Comment (mqasimsarfraz) [reply pkg/container-utils/cri/cri.go:356]: > we need to document and test container enrichers as we need to know exactly what elements is each enricher providing.

Totally agree, I know each runtime has subtle differences that we have to keep in mind so would be great to document it. 

Also, as part of this [issue](https://github.com/inspektor-gadget/inspektor-gadget/issues/1613) I [experimented](https://github.com/inspektor-gadget/inspektor-gadget/commit/e2c6158273116cd1a5bfa94d205ee23847220ccf) with `OCIConfig` for initial containers and it seemed to be [working fine](https://github.com/inspektor-gadget/inspektor-gadget/issues/1613#issuecomment-1540495562). Can help us unify use-cases for initial vs new containers.

File: pkg/container-collection/containers.go
Comment (mqasimsarfraz) [reply pkg/container-collection/containers.go:88]: Thanks for clarifying. I can only think of `BaseRuntimeMetadata` but I think it doesn't make much difference so I am fine keeping it as it is. Naming is hard topic :D

File: integration/inspektor-gadget/advise_networkpolicy_test.go
Comment (mqasimsarfraz) [reply integration/inspektor-gadget/advise_networkpolicy_test.go:91]: > It is because [ig-k8s doesn't directly use the Runtime enricher](https://github.com/inspektor-gadget/inspektor-gadget/blob/a39e7b00cc2d114259f3d9121aa4232ac7f6e407/pkg/gadgettracermanager/gadgettracermanager.go#L251-L320)

Make sense. Yeah, I see it is [used](https://github.com/inspektor-gadget/inspektor-gadget/blob/main/pkg/container-collection/k8s.go#L132) via Kubernetes client so it isn't clear if all the information will be available.

**Comment by mqasimsarfraz** [comment]
> --runtime.containername | -rc or --k8s.containername | -kc

I like this but we should keep `--containername` for backward compatibility for now. 

> But then what would we do in kubectl-gadget when we also make the runtimes columns available there?

I personally think a Kubernetes user might not care about runtime metadata compared to `ig` like `kubectl` doesn't have any columns but `crictl`, `ctr` and `docker` does use runtime container name. Also, this mean as a Kuberentes user I would expect `--containername` to be the one used in pod spec. So perhaps we shouldn't change anything for `kubectl-gadget`. What do you think?

File: pkg/container-utils/cri/cri.go
Comment (blanquicet) [reply pkg/container-utils/cri/cri.go:356]: That's great! Unifying initial vs new containers will help, but in general, we need an internal documentation saying how we are getting the information for each runtime and whether it works for both initial and new containers.

File: pkg/types/types.go
Comment (mauriciovasquezbernal) [inline pkg/types/types.go:151]: I'm going to write my comment down but we can handle this in another PR (I know we want to merge this soon to avoid having conflicts again). 

I think both `Runtime` and `K8s` should be pointers. The reason is that events can be generated outside Kubernetes and even outside containers, in that case it's more practical to have a nil pointer rather than an empty structure around.

File: pkg/types/types.go
Comment (mauriciovasquezbernal) [inline pkg/types/types.go:167]: I see that this is called from `EnrichEventByNetNs` for Kubernetes containers. The container runtime for all containers should be the same, so I think it's safe to set it even if we don't know the exact container within a pod.

File: pkg/operators/operators.go
Comment (mauriciovasquezbernal) [inline pkg/operators/operators.go:104]: I find it weird to have a `setContainerInfo` bool param in a `SetContainerMetadata` function. What about splitting this into `SetContainerMetadata` (as setContainerInfo=true) and `SetPodMetadata` (as setContainerInfo=false) and removing the boolean?

File: pkg/gadgettracermanager/gadgettracermanager.go
Comment (mauriciovasquezbernal) [inline pkg/gadgettracermanager/gadgettracermanager.go:179]: For another PR: We need to update the proto container definition to have the k8s and runtimes fields as our golang definition, otherwise this is going to be too confusing.

**Review by mauriciovasquezbernal — APPROVED** [review]
I have some minor comments. IMO it's good to go, great job!

(It's likely we missed something, it's very difficult to review, but we can fix that later on, I tested and it working fine, integration tests are passing, so not too much to worry about).

File: pkg/types/types.go
Comment (blanquicet) [reply pkg/types/types.go:104]: I prefer having specific types for the fields in the events so that people know what values to expect. Apart from the tests, this function is called in a few places, mainly to transform from flags to `RuntimeName` because in the core of the code all the functions use the `RuntimeName` type.

File: pkg/container-collection/options.go
Comment (blanquicet) [reply pkg/container-collection/options.go]: I know, but by the time I changed it to use `K8sMetadata`, the `RuntimeMetadata` doesn't exist yet so I can't go directly into the original commit.

File: pkg/types/types.go
Comment (blanquicet) [reply pkg/types/types.go:114]: That's a good idea. We don't even need to worry about people parsing it because, in that case, they will use the JSON format where it is divided.

File: pkg/types/types.go
Comment (blanquicet) [reply pkg/types/types.go:151]: I added it to the TODO list for future PRs. I will create a dedicated issue for them once this is merged.

File: pkg/types/types.go
Comment (blanquicet) [reply pkg/types/types.go:144]: That's in the TODO list for future PRs because it is tricky to enable `runtime` columns in `kubectl-gadget` without showing them by default as it is in `ig`. Vice versa in `ig` for `k8s` columns. In addition, I was considering adding a flag to display such info `ig <gadget> --k8s` and `kubectl-gadget <gadget> --runtime`. WDYT?

File: pkg/types/types.go
Comment (blanquicet) [reply pkg/types/types.go:187]: I added it to the todo list.

File: integration/ig/k8s/top_block_io_test.go
Comment (blanquicet) [reply integration/ig/k8s/top_block_io_test.go:40]: That's the standard Docker and CRI-O uses in k8s. Let's see if there is any case it is not like that and we can add consider it here. Right now, the test will fail in case it doesn't use such a prefix so we will immediately realize about it.

File: pkg/gadgettracermanager/gadgettracermanager.go
Comment (blanquicet) [reply pkg/gadgettracermanager/gadgettracermanager.go:179]: Added to the todo list.

File: pkg/operators/operators.go
Comment (blanquicet) [reply pkg/operators/operators.go:104]: Yep, it was also strange for me. Thanks for the suggestion. Applied.

File: pkg/types/types.go
Comment (blanquicet) [reply pkg/types/types.go:114]: I hid the Container ID but not yet the runtime. It is because I'd like to find an intelligent way to automatically add a virtual column (in this case, a column composed by `<runtimeName>://<containerName>) to all the structs that embed `BasicRuntimeMetadata`, instead of modifying the `GetColumns()` of all events to do so. Can I leave it for another PR?

cc @flyth

Comment (flyth) [reply pkg/types/types.go]: > I'd like to find an intelligent way to automatically add a virtual column (in this case, a column composed by <runtimeName>://<containerName>) to all the structs that embed BasicRuntimeMetadata, instead of modifying the GetColumns() of all events to do so.

Hmm, we could let such a type (like `BasicRuntimeMetadata`) implement an interface that we check for when initializing the columns (e.g. with a function like `InitColumns(columns)`) - where it then could add those virtual columns. But that might increase complexity quite a bit. I think with everything we need for the containerized gadgets in the columns lib, we should sooner or later make a plan with what features we need (in a more dynamic fashion), so we can identify overlaps and hopefully don't increase complexity too much.

**Pull request merged**
